runtime/src/installed_scheduler_pool.rs

```diff
@@ -0,0 +1,544 @@
+    type TransactionWithIndex<'tx>: WithTransactionAndIndex;
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct DefaultScheduleExecutionArg;
+
+impl ScheduleExecutionArg for DefaultScheduleExecutionArg {
+    type TransactionWithIndex<'tx> = &'tx (&'tx SanitizedTransaction, usize);
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum WaitReason {
+    // most normal termination waiting mode; couldn't be done implicitly inside Bank::freeze() -> () to return
+    // the result and timing in some way to higher-layer subsystems;
+    TerminatedToFreeze,
+    DroppedFromBankForks,
+    // scheduler will be restarted without being returned to pool in order to reuse it immediately.
+    ReinitializedForRecentBlockhash,
+}
+
+pub type InstalledSchedulerBox = Box<dyn InstalledScheduler<DefaultScheduleExecutionArg>>;
+
+#[derive(Clone, Debug)]
+pub struct SchedulingContext {
+    mode: SchedulingMode,
+    // Intentionally not using Weak<Bank> for performance reasons
+    bank: Arc<Bank>,
+}
+
+impl WithSchedulingMode for SchedulingContext {
+    fn mode(&self) -> SchedulingMode {
+        self.mode
+    }
+}
+
+impl SchedulingContext {
+    pub fn new(mode: SchedulingMode, bank: Arc<Bank>) -> Self {
+        Self { mode, bank }
+    }
+
+    pub fn bank(&self) -> &Arc<Bank> {
+        &self.bank
+    }
+
+    pub fn slot(&self) -> Slot {
+        self.bank().slot()
+    }
+
+    pub fn log_prefix(context: Option<&Self>, scheduler_id: SchedulerId) -> String {
+        const BITS_PER_HEX_DIGIT: usize = 4;
+
+        format!(
+            "sch_{:0width$x}{}: ",
+            scheduler_id,
+            context
+                .map(|c| format!("(slot:{}/mode:{:?})", c.slot(), c.mode()))
+                .unwrap_or_else(|| "(?)".into()),
+            width = SchedulerId::BITS as usize / BITS_PER_HEX_DIGIT,
+        )
+    }
+}
+
+// tiny wrapper to ensure to call wait_for_termination() via ::drop() inside
+// BankForks::set_root()'s pruning.
+#[derive(Clone)]
+pub struct BankWithScheduler {
+    inner: Arc<BankWithSchedulerInner>,
+}
+
+#[derive(Debug)]
+pub struct BankWithSchedulerInner {
+    bank: Arc<Bank>,
+    scheduler: InstalledSchedulerRwLock,
+}
+pub type InstalledSchedulerRwLock = RwLock<Option<InstalledSchedulerBox>>;
+
+#[allow(clippy::declare_interior_mutable_const)]
+pub const NO_INSTALLED_SCHEDULER_RW_LOCK: InstalledSchedulerRwLock = RwLock::new(None);
+
+impl BankWithScheduler {
+    pub(crate) fn new(bank: Arc<Bank>, scheduler: Option<InstalledSchedulerBox>) -> Self {
+        Self {
+            inner: Arc::new(BankWithSchedulerInner {
+                bank,
+                scheduler: RwLock::new(scheduler),
+            }),
+        }
+    }
+
+    #[cfg(any(test, feature = "test-in-workspace"))]
+    pub fn new_for_test(bank: Arc<Bank>, scheduler: Option<InstalledSchedulerBox>) -> Self {
+        Self::new(bank, scheduler)
+    }
+
+    pub fn new_without_scheduler(bank: Arc<Bank>) -> Self {
+        Self::new(bank, None)
+    }
```

Comment:
> Consider that BankWithScheduler::new(bank, None) is pretty much the same as BankWithScheduler::new_without_scheduler(bank).
> I do not think that the later makes the code any more readable than the former.

hmm, actually, I have different opinion. i think the `None` in `BankWithScheduler::new(bank, None)`  isn't clear as to what `None` refers to (`InstalledSchedulerBox` in this case). it could be some config struct or another auxiliary channel or like that. so i tend to define these small helper fns. Also, i thin it's ok for readers to read impl for clarity.

Reply:
If you do not like `None`, then go with option 1.
Require `InstalledSchedulerBox` in the `new()` call.

Reply:
forget to adjust option 1...: done 2e614f68d4d25d18aa97a81ad03787ea5773383d